agent/consul/leader_connect_ca.go

```diff
@@ -0,0 +1,1142 @@
+	newActiveRoot.IntermediateCerts = append(newActiveRoot.IntermediateCerts, intermediatePEM)
+	newActiveRoot.SigningKeyID = connect.EncodeSigningKeyID(intermediateCert.SubjectKeyId)
+
+	c.logger.Info("received new intermediate certificate from primary datacenter")
+	return nil
+}
+
+// intermediateCertRenewalWatch periodically attempts to renew the intermediate cert.
+func (c *CAManager) intermediateCertRenewalWatch(ctx context.Context) error {
+	isPrimary := c.serverConf.Datacenter == c.serverConf.PrimaryDatacenter
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(structs.IntermediateCertRenewInterval):
+			retryLoopBackoffAbortOnSuccess(ctx, func() error {
+				return c.RenewIntermediate(ctx, isPrimary)
+			}, func(err error) {
+				c.logger.Error("error renewing intermediate certs",
+					"routine", intermediateCertRenewWatchRoutineName,
+					"error", err,
+				)
+			})
+		}
+	}
+}
+
+// RenewIntermediate checks the intermediate cert for
+// expiration. If more than half the time a cert is valid has passed,
+// it will try to renew it.
+func (c *CAManager) RenewIntermediate(ctx context.Context, isPrimary bool) error {
+	// Grab the 'lock' right away so the provider/config can't be changed out while we check
+	// the intermediate.
+	if err := c.setState(CAStateRenewIntermediate, true); err != nil {
+		return err
+	}
+	defer c.setState(CAStateReady, false)
+
+	provider, _ := c.getCAProvider()
+	if provider == nil {
+		// this happens when leadership is being revoked and this go routine will be stopped
+		return nil
+	}
+	// If this isn't the primary, make sure the CA has been initialized.
+	if !isPrimary && !c.configuredSecondaryCA() {
+		return fmt.Errorf("secondary CA is not yet configured.")
+	}
+
+	state := c.delegate.State()
+	_, root, err := state.CARootActive(nil)
+	if err != nil {
+		return err
+	}
+	activeRoot := root.Clone()
+
+	// If this is the primary, check if this is a provider that uses an intermediate cert. If
+	// it isn't, we don't need to check for a renewal.
+	if isPrimary {
+		_, config, err := state.CAConfig(nil)
+		if err != nil {
+			return err
+		}
+
+		if _, ok := ca.PrimaryIntermediateProviders[config.Provider]; !ok {
+			return nil
+		}
+	}
+
+	activeIntermediate, err := provider.ActiveIntermediate()
+	if err != nil {
+		return err
+	}
+
+	if activeIntermediate == "" {
+		return fmt.Errorf("datacenter doesn't have an active intermediate.")
+	}
+
+	intermediateCert, err := connect.ParseCert(activeIntermediate)
+	if err != nil {
+		return fmt.Errorf("error parsing active intermediate cert: %v", err)
+	}
+
+	if lessThanHalfTimePassed(time.Now(), intermediateCert.NotBefore.Add(ca.CertificateTimeDriftBuffer),
+		intermediateCert.NotAfter) {
+		return nil
+	}
+
+	// Enough time has passed, go ahead with getting a new intermediate.
+	renewalFunc := c.getIntermediateCAPrimary
+	if !isPrimary {
+		renewalFunc = c.getIntermediateCASigned
+	}
+	errCh := make(chan error)
```

Comment:
```suggestion
	errCh := make(chan error, 1)
```
By making this a 1-buffered channel then you let each goroutine use the channel on its own timeline, rather than forcing them to wait on the other one. This is important when the context is canceled and the main goroutine leaves and stops receiving (forever). I think the Go GC will detect that nothing will ever read from the channel and stop the sending goutine, but best to make is more obvious how their lifecycles should go.